<?xml version="1.0"?>
<launch>
    <!-- world and robot -->
    <!-- arg for robots -->
    <arg name= "x" default="-2.0"/>
    <arg name ="y" default="-0.5"/>
    <arg name = "z" default="0.0"/>
    <arg name="roll" default="0.0"/>
    <arg name="pitch" default="0.0"/>
    <arg name="yaw" default="0.0"/>
    <arg name="no_static_map" default="false"/>

    <!--open world -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="world_name" value="$(find final_project)/worlds/costa_coffee_mod.world"/>
        <arg name="paused" value="false"/>
        <arg name="use_sim_time" value="true"/>
        <arg name="gui" value="true"/>
        <arg name="headless" value="false"/>
        <arg name="debug" value="false"/>
        <arg name="verbose" value="true"/>
    </include>
    <!-- put robot in world-->
    <include file="$(find final_project)/launch/put_robot_in_world.launch">
        <arg name="x" value="$(arg x)"/>
        <arg name="y" value="$(arg y)"/>
        <arg name="z" value="$(arg z)"/>
    </include>

  <!-- serving static map -->
  <node pkg="map_server" type="map_server" name="map_server" args="$(find final_project)/maps/costa_nav.yaml"/>
  <!-- AMCL.launch.xml -->
  <include file="$(find final_project)/config/amcl.launch.xml" />
  <!-- launch rviz -->
  <!-- YOU MUST CHANGE RVIZ FILE AS FINAL!!-->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find final_project)/rviz/costa_nav.rviz"/>



  <!-- ######################################### move base launch ############################################################# -->
  <arg name="cmd_vel_topic" default="/cmd_vel"/>
  <arg name="odom_topic" default="odom" />

  <arg name="base_global_planner" default="navfn/NavfnROS"/>
  <arg name="base_local_planner" default="dwa_local_planner/DWAPlannerROS"/>
<!-- <arg name="base_local_planner" default="base_local_planner/TrajectoryPlannerROS"/> -->


  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">

    <!-- Choose Planner!! You can use dwa planner or trajectory planner-->
    <param name="base_global_planner" value="$(arg base_global_planner)"/>
    <param name="base_local_planner" value="$(arg base_local_planner)"/>
    <rosparam file="$(find final_project)/params/move_base_params.yaml" command="load"/>

    <!-- observation sources located in costmap_common.yaml -->
    <rosparam file="$(find final_project)/params/costmap_common.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find final_project)/params/costmap_common.yaml" command="load" ns="local_costmap" />

    <!-- local costmap, needs size -->
    <rosparam file="$(find final_project)/params/costmap_local.yaml" command="load" ns="local_costmap" />
    <param name="local_costmap/width" value="2.0"/>
    <param name="local_costmap/height" value="2.0"/>

    <!-- static global costmap, static map provides size -->
    <rosparam file="$(find final_project)/params/costmap_global_static.yaml" command="load" ns="global_costmap" unless="$(arg no_static_map)"/>
    <param name="global_costmap/width" value="100"/>
    <param name="global_costmap/height" value="100"/>

    <!-- global costmap with laser, for odom_navigation_demo -->
    <rosparam file="$(find final_project)/params/costmap_global_laser.yaml" command="load" ns="global_costmap" if="$(arg no_static_map)"/>
    <rosparam file="$(find final_project)/params/move_base_params.yaml" command="load" />
    <rosparam file="$(find final_project)/params/dwa_local_planner_params.yaml" command="load"/>

    <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
    <remap from="odom" to="$(arg odom_topic)"/>

  </node>
  <!--####################################### WAYOINT #######################################-->
  <!--waypoints launch -->
  <env name="ROSCONSOLE_FORMAT" value="[${serverity}][${thread}][${node}/${function}:${line}]:${message}"/>
  <arg name="waypoints_topic" default="my_t3_waypoints"/>
  <node pkg='final_project' type='waypoint.py' name="moving_to_waypoints"/>
</launch>